error: '{' expected (got ";")

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: linux, Debian GNU/Linux 11 (bullseye)
Processor: 8 cpus, 64bit, little endian, Intel(R) Core(TM) i7-8565U CPU @ 1.80GHz
CC version: cc (Debian 10.2.1-6) 10.2.1 20210110

getwd: /home/premek/prog/chip8-v
vmodules: /home/premek/.vmodules
vroot: /home/premek/prog/v
vexe: /home/premek/prog/v/v
vexe mtime: 2022-04-13 22:01:40
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 a1e9cae.48c2951

Git version: git version 2.30.2
Git vroot status: weekly.2022.15-7-g48c29515
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 3654d6a8

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg hello.v`
```v
import datatypes

const (
	w = 64
	h = 32
)

interface Display {
mut:
	pixel(x int, y int, val bool)
	clear()
	refresh()
}

struct Vm {
mut:
	display Display
	stack   datatypes.Stack<u16>
}

struct Pattern {
	pattern []byte
	handler fn (mut m Vm)
}

fn new_pattern(pattern string, handler fn (mut m Vm)) Pattern {
	return Pattern{pattern.runes().map(byte('0x$it'.int())), handler}
}

fn main() {
	new_pattern('00E0', fn (mut m Vm) {
		println(m)
	})
}

```

**What did you expect to see?**

not 'This should never happen'


**What did you see instead?**
```
==================
/tmp/v_1000/hello.16893488910130859712.tmp.c:2027: warning: implicit declaration of function 'datatypes__Stack_T_u16_str'
/tmp/v_1000/hello.16893488910130859712.tmp.c:2027: error: '{' expected (got ";")
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/gen/c/auto_str_methods.v
+++ b/vlib/v/gen/c/auto_str_methods.v
@@ -896,7 +896,12 @@ fn (mut g Gen) gen_str_for_struct(info ast.Struct, styp string, str_fn_name stri
 		sftyp := g.typ(field.typ)
 		mut field_styp := sftyp.replace('*', '')
 		field_styp_fn_name := if sym_has_str_method {
-			'${field_styp}_str'
+			mut field_fn_name := '${field_styp}_str'
+			if sym.info is ast.Struct {
+				field_fn_name = g.generic_fn_name(sym.info.concrete_types, field_fn_name,
+					false)
+			}
+			field_fn_name
 		} else {
 			g.get_str_fn(field.typ)
 		}

--- a/vlib/v/tests/generics_struct_parent_has_str_to_string_test.v
+++ b/vlib/v/tests/generics_struct_parent_has_str_to_string_test.v
@@ -1,0 +1,36 @@
+import datatypes
+
+const (
+	w = 64
+	h = 32
+)
+
+interface Display {
+mut:
+	pixel(x int, y int, val bool)
+	clear()
+	refresh()
+}
+
+struct Vm {
+mut:
+	display Display
+	stack   datatypes.Stack<u16>
+}
+
+struct Pattern {
+	pattern []byte
+	handler fn (mut m Vm)
+}
+
+fn new_pattern(pattern string, handler fn (mut m Vm)) Pattern {
+	return Pattern{pattern.runes().map(byte('0x$it'.int())), handler}
+}
+
+fn test_generics_struct_parent_has_str_to_string() {
+	p := new_pattern('00E0', fn (mut m Vm) {
+		println(m)
+	})
+	println(p)
+	assert true
+}

